web3/providers/base.py

```diff
@@ -60,12 +60,25 @@ def decode_rpc_response(self, response):
         return json.loads(to_text(response))
 
     def encode_rpc_request(self, method, params):
-        return to_bytes(text=json.dumps({
-            "jsonrpc": "2.0",
-            "method": method,
-            "params": params or [],
-            "id": next(self.request_counter),
-        }))
+        rpc_id = next(self.request_counter)
+        rpc_params = params or []
+        try:
+            j_str = json.dumps({"jsonrpc": "2.0",
+                                "method": method,
+                                "params": rpc_params,
+                                "id": rpc_id})
+        except TypeError:
+            err_msg = ''
+            for i, p in enumerate(params):
+                try:
+                    json.dumps(p)
+                except TypeError:
+                    err_msg += ' ({},{});'.format(i, repr(p))
+            msg = 'encode_rpc_request failure for method {} for invalid (index, parameter): '
```

Comment:
Instead of embedding this functionality within the provider, this strikes me as something that could be extracted to a stand-alone function and tested much easier in isolation.

```python
def json_encode_with_helpful_error_message(payload):
    try:
        return json.dumps(payload)
    except:
        ...  # awesome error messaging code.
```

Thoughts?  Should help cleanup this method body a lot and lower the overhead to testing.